File: src/vm/src/lexer.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include ".hs/lexer.h"

Vm vm_init() {
	Vm vm = {0};
	return vm;
}

void get_tok(Vm *vm, char **src) {
	if (src[vm->index] == NULL) {
		vm->cur_tok = STR_END;
		return;
	}
	
	while (strcmp(src[vm->index], " ") == 0) {
        vm->index++;
    }

	printf("%s\n", src[vm->index]);

	if (strcmp(src[vm->index], "PUSH") == 0) {
		vm->cur_tok = PUSH;
	} else if (strcmp(src[vm->index], "BIND") == 0) {
		vm->cur_tok = BIND;
	} else {
		if (src[vm->index][0] >= '0' && src[vm->index][0] <= '9') {
			vm->cur_tok = NUMBER;
		} else {
			vm->cur_tok = ID;
		}
	}
	vm->index++;
}

File: src/vm/src/lexer.rs
#[derive(Debug)]
pub enum Token {
    PUSH,
    NUMBER(i32),
    BIND,
    ID(String),
    EOF,
}

pub struct Vm {
    pub index: usize,
    pub cur_tok: Token,
    pub stack: Vec<i32>,
    pub cur_line: i64,
}

// Function that inits the vm
pub fn lex_init() -> Vm {
    Vm {
        index: 0,
        cur_tok: Token::EOF,
        stack: Vec::new(),
        cur_line: 0,
    }
}

// Main lexer function that updates the vm state
pub fn get_tok(vm: &mut Vm, src: &Vec<String>) {
    let c = &src[vm.index];

    match c.as_str() {
        " " => get_tok(vm, src),
        "PUSH" => vm.cur_tok = Token::PUSH,
        "BIND" => vm.cur_tok = Token::BIND,
        &_ => {
            if c.parse::<i32>().is_ok() {
                vm.cur_tok = Token::NUMBER(
                    c.parse::<i32>()
                        .expect("This should not happen report this"),
                );
            } else {
                vm.cur_tok = Token::ID(String::from(c));
            }
        }
    }
    vm.index += 1;
}
